### DISCUSSION TITLE
How to set value of nested environment variable in `cypress.json` file?

### DISCUSSION THREAD

**Discussion by Moataz-Mahmoud** [cypress-io/cypress#14374] [Questions and Help] [answered]
**The Problem:**
My config file has hundreds of environment variables which makes it unusable. So I thought to use nested environment variables to organize things.
I'm able to get the values of a nested variable easily, but I struggle to set its value **without clearing the value of the other variables in the same block**

**Here is an example:** 

`cypress.json`

```
{
  env: {
    "shortcodes": {
      "dedicated_shortcode": "",
      "shared_shortcode": "",
      "region": "US",
      "description": "automated dedicated shortcode",
      "shortcode_id": ""
    }
  }
}
```

I set the values in a `before` block like this:

```
Cypress.env(
      {
        shortcodes:
        {
          token: Cypress.env('shortcodes').token,
          region: Cypress.env('shortcodes').region,
          description: Cypress.env('shortcodes').description,
          dedicated_shortcode: (Math.floor(10000 + Math.random() * 90000).toString()),
          shared_shortcode: (Math.floor(10000 + Math.random() * 90000).toString()),
        },
      },
    );
```
When I try to reassign value to any of the values inside this `shortcodes` block, I have to state the values of the other fields.
For example, if I did something like that `Cypress.env({ shortcodes: { shortcode_id: response.body.id } });` it clears all the values of the other entities in `shortcodes` block and assign new value for only the `shortcode_id`.

Is there a notation which I can set the value of one of the entities without resetting the others or stating them explicitly?

**Comment by bahmutov** [comment, marked as answer]
I think you are making it more complex. Just do
```js
const env = Cypress.env()
env.shortcodes.dedicated_shortcode = Math.floor(10000 + Math.random() * 90000).toString()
env.shortcodes.shared_shortcode = Math.floor(10000 + Math.random() * 90000).toString()
```

You don't even need to call `Cypress.env(env)` after that - since it is the same object reference anyway.
To confirm, see this recipe we run to confirm it https://glebbahmutov.com/cypress-examples/6.2.0/recipes/set-env-prop.html

**Reply by myildizCH** [reply]
@bahmutov When we have nested JSON in our cypress.env.json, how should we set the env variables in CI? For ex: 

In cypress.env.json:

```
{
"dev": {
    "username": "Mike"
   },

"staging": {
    "username": "Lara"
   }
}
```

Will we enter env variables in CI tool like `CYPRESS_DEV.USERNAME = Mike` ?

**Comment by coler-j** [comment]
How do you provide typing (i.e. in Typescript) for nested env types?
